### DISCUSSION TITLE
How to replicate `dr.cuda.ArrayXf([1,2,3])` using Mitsuba?

### DISCUSSION THREAD

**Discussion by maxfrei750** [mitsuba-renderer/mitsuba3#679] [Q&A] [answered]
Hi! I'd like to create the same output as the abovementioned command, but using Mitsuba, so that the data type depends on the user-selected backend. How can I do that?

Also: Is there documentation for this really basic vector/array/tensor stuff? I tried to look in the mitsuba and drjit docs, but didn't find a basic explanation on how to use the different data structures.

Thanks!

**Comment by njroussel** [comment, marked as answer]
Hi @maxfrei750  !!

It looks like `ArrayXf` was never bound/aliased in Mitsuba. I wonder if this was intentional or if we just never needed it (you can usually use `TensorXf` to accomplish the same goal). Until I figure that out, you should be able to at least monkey patch it with Python.

We definitely still need to and want to improve the documentation: Tensors and textures aren't even documented in Dr.Jit. Is there something specific that you feel like is lacking? "basic vector/array/tensor stuff" what were you expecting?

**Reply by maxfrei750** [reply]
Hi @njroussel! Thanks for the swift reply and the hint with `TensorXf`. It tried it before, but unfortunately, I didn't manage to achieve the same result with respect to the shape: 

```
>> dr.cuda.ArrayXf([1,2,3]).Shape
(18446744073709551615, 18446744073709551615)

>> mi.TensorXf([1,2,3],shape=(3,)).Shape
(18446744073709551615,)
```

That behavior is, for instance, along the "stuff", I was referring to. :smile: I guess that the explanation is easy, but the "open" dimensions of the `drjit` data types are still confusing to me, coming from Numpy and PyTorch.

Generally, what are the available types of data structures, how to create them, how are they related, what makes them special, as well as when and how to use them. Maybe some comparisons to Numpy, because most people will be somewhat familiar with it. Starting with stuff so basic that people like me would probably be too ashamed to ask for them and figure them out for themselves by trial and error. :sweat_smile: Basically just a bunch of examples, with a little explanation. 

Thanks once again!

**Reply by maxfrei750** [reply]
@njroussel For the time being, do you have any advice on how to realize an equivalent output using `mi.TensorXf`? Thanks!

**Reply by njroussel** [reply]
Thanks for the feedback!

I'm a bit surprised that you're using `.Shape` :exploding_head:. This is mostly meant for internal uses, whereas `dr.shape(var)` or `var.shape` is what you'd usually want. Can I ask what you're trying to use it for? The `Shape` of a `TensorXf` will indeed always be uni-dimensional.

**Reply by maxfrei750** [reply]
Thanks for the hint! That's so helpful. The reason, I went for `.Shape` is rather dumb: it's what shows up in VSCode's debugger, and I wasn't aware of `dr.shape()`. :facepalm: 

`var.shape` doesn't work for me:
```
>> dr.cuda.ArrayXf(1,2,3).shape
AttributeError: 'drjit.cuda.ArrayXf' object has no attribute 'shape'
```
Was that property added recently? Then maybe I need to update... It would be a nice feature, because it is more convenient and would also show up in the debugger.

Using `dr.shape`, I was able to get a much better understanding of my problem, which might be a little much to unpack here (or might even be a bug), so I'll open another discussion, to keep things tidy. Thanks a ton and see you in the next discussion. :sweat_smile:

**Reply by njroussel** [reply]
No, it's always been there I think :sweat_smile: 

We use a script to automatically generate files that are use by your IDE's auto-complete system. It's very likely this doesn't wrotk that well.

**Reply by maxfrei750** [reply]
> No, it's always been there I think sweat_smile

I think there was a misunderstanding on my behalf. When referring to `var.shape`, you were exclusively talking about `mi.TensorXf` (which indeed has the `shape` attribute), correct ? I thought that it was also supposed to work for e.g. `Float` and `Array` types (since `dr.shape` can be used on them).

That being said, IMHO, it would be convenient to have the `shape` attribute for the other types as well.
